### PR TITLE
[JENKINS-65790]: Remove injected styles and scripts in form validation

### PR DESCRIPTION
**Pull request by sabberworm**
See [JENKINS-65790](https://issues.jenkins-ci.org/browse/JENKINS-65790).

In case of errors during form validation, the response was used verbatim. Improves this behaviour somewhat by stripping styles and scripts from the response. Also uses a `<details>` element instead of a click handler on an `<a>` tag to show/hide the error response.

### Proposed changelog entries

N/A

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@daniel-beck

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/5601"></a>

### CODE REVIEW

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (sabberworm) [inline war/src/main/webapp/scripts/hudson-behavior.js:2927]: We should also check that the response is in fact HTML and not some other Content-Type. The original code also didn’t do this so I’m unsure if there are any endpoints that don’t respond with HTML in some cases.

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (sabberworm) [inline war/src/main/webapp/scripts/hudson-behavior.js:2928]: I couldn’t test this special-cased element as I don’t know how to trigger it. Any hints appreciated.
The original implementation didn’t clearly specify that the `error-description` element is in the returned part of the DOM and not anywhere else but it’s the only way the code made sense (otherwise, why inject the elements only to remove them again?), so that’s what I assumed.

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (sabberworm) [inline war/src/main/webapp/scripts/hudson-behavior.js:2940]: I did opt to put in some more text to indicate what kind of error we’re dealing with. Not sure if i18n is required (or how that would be done).

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (sabberworm) [inline war/src/main/webapp/scripts/hudson-behavior.js:2943]: This is how it looks:

image


And opened:

image

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (sabberworm) [reply war/src/main/webapp/scripts/hudson-behavior.js:2940]: I now realize that “Field Validation ERROR” can be assumed to just be the same as a field that doesn’t match the validation constraints.

Maybe this is clearer:

```suggestion
        summary.textContent = 'Error validating this field';
```

**Review by miglen — APPROVED** [review]

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (daniel-beck) [reply war/src/main/webapp/scripts/hudson-behavior.js:2928]: This is from `oops.jelly`; it is a hack I did a few years ago mostly that "Apply" buttons get a useful error message without including an entire layout in their popup-looking dialog.

Might still be triggerable when trying to Apply a job configuration with a syntactically invalid "Build periodically" cron pattern.

**Comment by alecharp** [comment]
@daniel-beck @Wadeck do you have any concerns about the content of this PR?

**Comment by basil** [comment]
Needs a review from @jenkinsci-cert.

**Comment by timja** [comment]
The linked Jira doesn't seem to be related to this.

Also, form validation had some re-work done recently so this is now conflicted

**Comment by sabberworm** [comment]
> The linked Jira doesn't seem to be related to this.

In what way? The linked Jira is caused by a global style from the default sinatra error page being injected into the page. The change I propose would ensure no styles get injected.

**Review by Kevin-CB — APPROVED** [review]
We apologize for the delay, we reviewed your PR and there is no blocking from security team at this point.

**Comment by NotMyFault** [comment]
While we have the go from the security team for the proposed change, a more contemporary diff has been implemented in the meantime to validate form input.
If the jira issue applies to the new form validation still, the PR needs to be revised to adapt the recent changes, because the current proposal is no longer applicable.

**Comment by Wadeck** [comment]
_(side topic)_
@NotMyFault It could be interesting to find a way to prevent more recent initiatives to make previous PRs irrelevant, instead of re-using / adapting them to suit the new goal. That could reduce frustration from non-very-active contributors who's effort should also be valued. Any idea could help ;) 
Topic added for the next UI/UX SIG meeting

**Comment by NotMyFault** [comment]
> That could reduce frustration from non-very-active contributors who's effort should also be valued. Any idea could help ;)

Yeah, that's an understandable and actually relatable concern. Our backlog of PRs decreased a lot (thanks to everyone triaging, reviewing and integrating PRs 👏🏻), but basil's review request of a security team member basically brought this PR back to life.

To be fair, keeping such a small PR (diff wise) on hold for almost a year without feedback the submitter can address doesn't facilitate a fast review process or a smooth integration without conflicts at all.

**Comment by basil** [comment]
> Who’s to say that, after me fixing the conflict right now, the PR won’t remain open for another year until someone does the next conflicting change, at which point I will be asked to rework again?

You are right, and on behalf of the core team I apologize for this. We will try to do better in the future. I have been making a concerted effort to shepherd recent PRs across the finish line in a timely fashion and to address the PR backlog. I am grateful to my colleagues on the Jenkins core and security teams for their valued coöperation in this regard.

> If the jira issue applies to the new form validation still, the PR needs to be revised to adapt the recent changes, because the current proposal is no longer applicable.

So does the Jira issue apply to the new form validation, or does it not? After reading this thread carefully the answer to this question remains unclear to me. Based on [JENKINS-65790 (comment)](https://issues.jenkins.io/browse/JENKINS-65790?focusedCommentId=410487&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-410487) it seems that the issue described in [JENKINS-65790](https://issues.jenkins.io/browse/JENKINS-65790) only applies when deprecated plugins are installed, and I have not seen any indication that the issue can be reproduced on an installation without any deprecated plugins. If the issue can still be reproduced on the latest weekly without any deprecated plugins installed, then the issue remains valid and we can discuss the next steps for the implementation in this PR. If the issue cannot be reproduced on the latest weekly without any deprecated plugins installed, then in approximately a week I plan to close both the issue (as "Cannot Reproduce") and this PR.

**Comment by timja** [comment]
Closing as there's no diff remaining in this PR and no update in awhile. Apologies for the PR being missed earlier

**Pull request closed without merging**
